src/mlpack/methods/xgboost/loss_functions/sse_loss.hpp

```diff
@@ -86,22 +76,114 @@ class SSELoss
   template<bool UseWeights, typename MatType, typename WeightVecType>
   double Evaluate(const MatType& input, const WeightVecType& /* weights */)
   {
+    // Sanity check for empty input.
+    if (input.n_cols == 0)
+      return 0.0;
+
     // Calculate gradients and hessians.
     gradients = (input.row(1) - input.row(0)).t();
     hessians = arma::vec(input.n_cols, arma::fill::ones);
 
-    return std::pow(ApplyL1(arma::accu(gradients)), 2) /
-        (arma::accu(hessians) + lambda);
+    // Calculates total sum of gradients and hessians.
+    gTotal = arma::accu(gradients);
+    hTotal = arma::accu(hessians);
+
+    return std::pow(ApplyL1(gTotal), 2) / (hTotal + lambda);
+  }
+
+  /**
+   * Sorts the gradients and hessians according to the sorted order of the
+   * feature.
+   *
+   * @param sortedIndices Vector of indices according to the sorted order of
+   *      the feature.
+   * @param sortedGradients Vector to store the sorted gradients.
+   * @param sortedHessians Vector to store the sorted hessians.
+   */
+  void sortGradAndHess(const arma::uvec& sortedIndices,
+                       arma::vec& sortedGradients,
+                       arma::vec& sortedHessians)
+  {
+    for (size_t i = 0; i < sortedIndices.n_elem; ++i)
+    {
+      sortedGradients[i] = gradients[sortedIndices[i]];
+      sortedHessians[i] = hessians[sortedIndices[i]];
+    }
```

Comment:
It's possible this could be made faster via a one-liner:

```
sortedGradients = gradients[sortedIndices];
sortedHessians = hessians[sortedIndices];
```

But, that is so simple that perhaps it's better to just inline that where you need it instead of having this function.  Your call. :)

Reply:
We can't do that inline, because `gradients` and `hessians` not accessible from outside of `LossFunction`. I can make the function itself an `inline` function.

Reply:
Do you mean that this function is called from `XGBExactNumericSplit` and not internally from `SSELoss`?  I disagree with that design: this now means there is another point of linkage between `XGBExactNumericSplit` and `SSELoss`, and another function that someone would need to implement in order to implement a new loss function.

Are you really sure this is necessary?  Can we not have higher levels provide a pre-sorted set of gradients and hessians to `BinaryScanInitialize()` or `Evaluate()`?

Reply:
I am not sure if we can do that. The reason is that we need to sort the gradients again and again for every split because we are sorting them based on the sorted order of the data. So, we have to sort them each time. Now, we have to sort them before calling `BinaryScanInitialize()` because of [this](https://github.com/RishabhGarg108/mlpack-1/blob/83ee7b1dc954cb63a936e9d7e0f4f54b77f3e718/src/mlpack/methods/xgboost/xgb_exact_numeric_split_impl.hpp#L44) check.

Thus it can't be possible as long as we are storing gradients in the `SSELoss`. Now one way we can do is by having a `BaseLoss` class and implement it there and then all the loss functions can inherit from it. This will reduce the code significantly because if you will take a closer look then you will see that there is nothing within the `BinaryScanInitialize()`, `BinaryStep()` and `BinaryGains()` which is specific to `SSELoss`. We can also move `OutputValue()` to the base class. The reason that these four functions are independent of the choice of loss function because all we are doing is working with gradients and hessians in these. So, they will always be the same for all loss functions.

Do you think this will be a good idea? It will definitely reduce the number of functions that needs to be implemented for sure.

Reply:
Why don't we represent the gradients and hessians as part of the responses?  That general idea should be adaptable to solve the issue.  I thought we had discussed this concept before, so hopefully the idea should be familiar.

Let us not have a base class for the losses.  Templates don't necessarily require a base class; we just need to be careful about the signatures of everything we are implementing.